radix-engine/src/ledger/traits.rs

```diff
@@ -33,110 +33,78 @@ pub trait QueryableSubstateStore {
 
 /// A ledger stores all transactions and substates.
 pub trait SubstateStore {
-    /// Top Level Objects
-    fn get_resource_def(&self, resource_def_id: ResourceDefId) -> Option<ResourceDef>;
-    fn put_resource_def(&mut self, resource_def_id: ResourceDefId, resource_def: ResourceDef);
-    fn get_package(&self, package_id: PackageId) -> Option<Package>;
-    fn put_package(&mut self, package_id: PackageId, package: Package);
-    fn get_component(&self, component_id: ComponentId) -> Option<Component>;
-    fn put_component(&mut self, component_id: ComponentId, component: Component);
-
-    /// Child Objects
-    fn get_lazy_map_entry(
-        &self,
-        component_id: ComponentId,
-        lazy_map_id: &LazyMapId,
-        key: &[u8],
-    ) -> Option<Vec<u8>>;
-    fn put_lazy_map_entry(
-        &mut self,
-        component_id: ComponentId,
-        lazy_map_id: LazyMapId,
-        key: Vec<u8>,
-        value: Vec<u8>,
-    );
-    fn get_vault(&self, component_id: ComponentId, vault_id: &VaultId) -> Vault;
-    fn put_vault(&mut self, component_id: ComponentId, vault_id: VaultId, vault: Vault);
-    fn get_non_fungible(
-        &self,
-        resource_def_id: ResourceDefId,
-        key: &NonFungibleKey,
-    ) -> Option<NonFungible>;
-    fn put_non_fungible(
-        &mut self,
-        resource_def_id: ResourceDefId,
-        key: &NonFungibleKey,
-        non_fungible: NonFungible,
-    );
+    fn get_substate<T: Encode>(&self, address: &T) -> Option<Vec<u8>>;
+    fn get_child_substate<T: Encode>(&self, address: &T, key: &[u8]) -> Option<Vec<u8>>;
+
+    fn put_substate<T: Encode>(&mut self, address: &T, substate: &[u8]);
+    fn put_child_substate<T: Encode>(&mut self, address: &T, key: &[u8], substate: &[u8]);
 
     fn bootstrap(&mut self) {
-        if self.get_package(SYSTEM_PACKAGE).is_none() {
+        let package: Option<Package> = self
+            .get_substate(&SYSTEM_PACKAGE)
+            .map(|v| scrypto_decode(&v).unwrap());
```

Comment:
`scrypto_decode` and `scrypto_encode` everywhere are really annoying me. Can we have **default methods** in the `SubstateStore` trait that accepts rust structs and automatically decodes/encodes?

Reply:
This works for now.

Reply:
Agreed. This is why I proposed an independent layer (on top of substates) earlier. Didn't want to bring in too much extra work into this PR, so proposed default methods.